File: public/dsa/two_pointers/three_sum/pointers/three_sum.py
from typing import List


class Solution:
    def threeSum(self, A: List[int]) -> List[List[int]]:
        n = len(A)
        A.sort()
        result = []
        for i in range(n):
            if i > 0 and A[i] == A[i-1]:
                continue
            left = i + 1
            right = n - 1
            while left < right:
                if left > i + 1 and A[left] == A[left-1]:
                    left += 1
                    continue
                triplet_sum = A[i] + A[left] + A[right]
                if triplet_sum > 0:
                    right -= 1
                elif triplet_sum < 0:
                    left += 1
                else:
                    result.append([A[i], A[left], A[right]])
                    left += 1
                    right -= 1
        return result


# tests-start
A = [1, 1, 0, -1, -2]
sol = Solution()
print(sol.threeSum(A))

# tests-end

'''output
[[-2, 1, 1], [-1, 0, 1]]
'''

File: public/dsa/two_pointers/three_sum/pointers/three_sum.cpp
#include <bits/stdc++.h>
using namespace std;

#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> threeSum(vector<int>& A) {
    int n = A.size();
    sort(A.begin(), A.end());
    vector<vector<int>> result;

    for (int i = 0; i < n; i++) {
        if (i > 0 && A[i] == A[i - 1])
            continue;

        int left = i + 1;
        int right = n - 1;

        while (left < right) {
            if (left > i + 1 && A[left] == A[left - 1]) {
                left++;
                continue;
            }

            int triplet_sum = A[i] + A[left] + A[right];

            if (triplet_sum > 0) {
                right--;
            } else if (triplet_sum < 0) {
                left++;
            } else {
                result.push_back({A[i], A[left], A[right]});
                left++;
                right--;
            }
        }
    }

    return result;
}

// tests-start

int main() {
    vector<int> A = {1, 1, 0, -1, -2};
    vector<vector<int>> result = threeSum(A);

    for (const auto& triplet : result) {
        cout << "[";
        for (int i = 0; i < triplet.size(); ++i) {
            cout << triplet[i];
            if (i < triplet.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
    }

    return 0;
}

// tests-end

/*output
[-2, 1, 1]
[-1, 0, 1]
*/